openedx/openedx-platform
# reafactor: simplify XModule serialization/deserialization layer [BD-13]

## Description

Most of the methods in `XmlMixin` act as wrappers for the official API for serialization and deserialization (`parse_xml()` and `add_xml_to_node()`). `XmlParserMixin` contains the code which does the actual serialization and deserialization.

This simplifies the serialization layer by:
1. Replacing remaining old `from_xml` deserialization methods with `parse_xml`, which is used by XBlocks.
2. Removing the `XmlMixin` class and renaming `XmlParserMixin` to `XmlMixin`. This way, the code handling the serialization and deserialization is retained.
3. Removing the `XmlDescriptor` class, which was used only in tests.
4. Removing `slug` from `url_name` lookup, as we had already removed its translation in https://github.com/openedx/edx-platform/pull/29927.

## Sandbox

LMS: https://pr30072.sandbox.opencraft.hosting/
Studio: https://studio.pr30072.sandbox.opencraft.hosting/

The sandbox contains one manual change because the Error XBlock is not loaded by default when importing a course. [This line](https://github.com/openedx/edx-platform/blob/a389a9ff10fd06bd053107fc6a4f19c7d24f7233/xmodule/modulestore/xml.py#L332) is replaced with:
```diff
---         self.load_error_modules = load_error_modules
+++         self.load_error_modules = True
```


## Testing instructions

1. Log in as a `staff` user.
1. Open any sample course or create a new course. Go to Tools > Import and upload file [good.tar.gz](https://github.com/openedx/edx-platform/files/8098012/good.tar.gz)
1. Once the file is imported, verify that all XBlocks were loaded correctly
1. Go to Tools > Export and download the file.
1. Verify that the export's structure is correct.
1. Go to Tools > Import again, and upload file [bad.tar.gz](https://github.com/openedx/edx-platform/files/8098030/bad.tar.gz).
1. In this file, the CAPA module is malformed. Verify that the Error block is rendered in place of the CAPA Multiple choice block.

## Other information

`XmlParserMixin` (renamed to `XmlMixin` now)  and some XBlock classes will continue to retain their custom logic. This logic cannot be removed without breaking backward compatibility. Since this logic does not impact the complexity of the system there are no negatives to retaining it.

Private-ref: [BB-5573](https://tasks.opencraft.com/browse/BB-5573)
